src/types/array-node.ts

```diff
@@ -100,15 +100,23 @@ export class ArrayNode extends Node {
 }
 
 export function createArrayFactory<S, T extends S>(subtype: IModelFactory<S, T>): IModelFactory<S[], IObservableArray<T>> {
-    return createFactory(
+    let factory = createFactory(
         "array-factory",
-        ArrayNode,
-        {
-            subType: subtype,
-            isArrayFactory: true
-        } as IArrayFactoryConfig,
-        () => observable.shallowArray()
+        "array",
+        snapshot => (Array.isArray(snapshot) && snapshot.every(item => subtype.is(item))),
```

Comment:
@mattiamanzati so you want to be able to do `unionOf(arrayOf(X), arrayOf(Y))`? But that still cannot pick the correct type for an empty snapshot I would guess, so maybe the type system should not be _that_ flexible, and only allow unions of model types?